scripts/documentation/tsdoc.ts

```diff
@@ -0,0 +1,356 @@
+    try {
+      const relativePathToReadme = path.resolve('packages', directoryPath, README_FILE);
+      if (fs.existsSync(relativePathToReadme)) {
+        return directoryPath;
+      }
+      // look one directory level up
+      return this.getFilePathName(directoryPath);
+    } catch (err) {
+      console.error(err); //tslint:disable-line
+    }
+    return '';
+  }
+
+  /**
+   * Generates Markdown file for each entry in `this.markdownBuffer`,
+   * which is populated from `this.generateDocsForModule()`.
+   * @param markdownBuffer object where keys are the module name, and
+   * value is the module documentation in json format, ready to be
+   * formatted into markdown.
+   */
+  async generateMarkdownFileFromBuffer(markdownBuffer: MarkdownBuffer) {
+    for (const readmeDirectoryPath in markdownBuffer) {
+      /**
+       * This currently only has been tested on mdc-drawer.
+       * TODO: remove this if condition once all READMEs are generated
+       */
+      const allowList = [
+        'mdc-drawer',
+      ];
+
+      if (allowList.some((allowed) => readmeDirectoryPath.includes(allowed))) {
+        const readmeDestinationPath = `./packages/${readmeDirectoryPath}/${README_FILE}`;
+        const finalReadmeMarkdown = await this.insertMethodDescriptionTable(markdownBuffer, readmeDirectoryPath);
+        fs.writeFile(readmeDestinationPath, finalReadmeMarkdown, (error) => {
+          console.log(`~~ generated ${readmeDestinationPath}`); // tslint:disable-line
+          if (error) {
+            console.error('error ', error); //tslint:disable-line
+          }
+        });
+      }
+    }
+  }
+
+  /**
+   * Returns a promise, that resolves with the finalized markdown with
+   * inserted documentation in markdown table format.
+   * @param markdownBuffer object where keys are the module name, and
+   * value is the module documentation in json format, ready to be
+   * formatted into markdown.
+   * @param readmeDirectoryPath directory path to readme file
+   * (ie. mdc-textfield/character-counter or mdc-drawer)
+   * @return Promise<{string}>
+   */
+  private async insertMethodDescriptionTable(
+    markdownBuffer: MarkdownBuffer,
+    readmeDirectoryPath: string,
+  ) {
+    const readmeMarkdownPath = `./packages/${readmeDirectoryPath}/${README_FILE}`;
+    const readmeMd = await readFile(readmeMarkdownPath, 'utf8');
+    const componentModules = markdownBuffer[readmeDirectoryPath].component
+      .filter((module) => module.methods.length || module.properties.length || module.events.length)
+      .sort(this.sortByModuleType);
+    const nonComponentModules = markdownBuffer[readmeDirectoryPath].nonComponent
+      .filter((module) => module.methods.length || module.properties.length || module.events.length)
+      .sort(this.sortByModuleType);
+    const apiMarkdownTable =
+      this.templateFunction({modules: componentModules, showFrameworkUsage: true}) +
```

Comment:
wouldn't it be easier if template function takes all template data it needs at once?

```ts
this.templateFunction({
  componentModules,
  nonComponentModules,
});
```

Reply:
Well it is the same template, we'd create another template (partial) to combine these together. I don't think it matters either way works. What would you like?

Reply:
I think having a single template as a whole makes it easier to reformat later. The current implementation we are not fully leveraging the use of templating engine, we concat two strings generated separately in the code instead.

Reply:
Ok sounds good - i'll make the change.

Reply:
Updated!